File: 2_year_2_semester/spo/football-championship-system/backend/app/schemas/player.py
from typing import Optional
from pydantic import BaseModel, validator
from datetime import datetime

class PlayerBase(BaseModel):
    first_name: str
    last_name: str
    age: int
    jersey_number: int
    position: str
    
    @validator('age')
    def validate_age(cls, v):
        if v < 16 or v > 45:
            raise ValueError('Age must be between 16 and 45')
        return v
    
    @validator('jersey_number')
    def validate_jersey_number(cls, v):
        if v < 1 or v > 99:
            raise ValueError('Jersey number must be between 1 and 99')
        return v
    
    @validator('position')
    def validate_position(cls, v):
        valid_positions = ['GK', 'DEF', 'MID', 'FWD']
        if v not in valid_positions:
            raise ValueError(f'Position must be one of {valid_positions}')
        return v

class PlayerCreate(PlayerBase):
    team_id: int

class PlayerUpdate(BaseModel):
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    age: Optional[int] = None
    jersey_number: Optional[int] = None
    position: Optional[str] = None
    team_id: Optional[int] = None
    
    @validator('age')
    def validate_age(cls, v):
        if v is not None and (v < 16 or v > 45):
            raise ValueError('Age must be between 16 and 45')
        return v
    
    @validator('jersey_number')
    def validate_jersey_number(cls, v):
        if v is not None and (v < 1 or v > 99):
            raise ValueError('Jersey number must be between 1 and 99')
        return v
    
    @validator('position')
    def validate_position(cls, v):
        if v is not None:
            valid_positions = ['GK', 'DEF', 'MID', 'FWD']
            if v not in valid_positions:
                raise ValueError(f'Position must be one of {valid_positions}')
        return v

class PlayerInDBBase(PlayerBase):
    id: int
    team_id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        orm_mode = True

class Player(PlayerInDBBase):
    pass

class PlayerWithTeam(PlayerInDBBase):
    team: 'Team'
    
    class Config:
        orm_mode = True

# Avoid circular imports
from app.schemas.team import Team
PlayerWithTeam.update_forward_refs()

File: 2_year_2_semester/spo/football-championship-system/frontend/src/types/player.ts
// frontend/src/types/player.ts
import { Team } from './team';

/**
 * Позиции игроков, используемые на бэкенде.
 */
export type PlayerPosition = 'GK' | 'DEF' | 'MID' | 'FWD';

/**
 * Базовый интерфейс для игрока.
 */
export interface Player {
  id: number;
  team_id: number;
  first_name: string;
  last_name: string;
  age: number;
  jersey_number: number;
  position: PlayerPosition;
  created_at: string;
  updated_at: string;
}

/**
 * Интерфейс для игрока с детальной информацией о команде.
 * Соответствует схеме `PlayerWithTeam` в Pydantic.
 */
export interface PlayerWithTeam extends Player {
  team: Team;
}

/**
 * Интерфейс для создания нового игрока.
 * Соответствует схеме `PlayerCreate` в Pydantic.
 */
export interface PlayerCreate {
  first_name: string;
  last_name: string;
  age: number;
  jersey_number: number;
  position: PlayerPosition;
  team_id: number;
}

/**
 * Интерфейс для обновления существующего игрока.
 * Поля необязательные, соответствует схеме `PlayerUpdate` в Pydantic.
 */
export interface PlayerUpdate {
  first_name?: string;
  last_name?: string;
  age?: number;
  jersey_number?: number;
  position?: PlayerPosition;
  team_id?: number;
}

/**
 * Интерфейс для параметров фильтрации списка игроков.
 */
export interface PlayerListParams {
    skip?: number;
    limit?: number;
    team_id?: number;
    position?: PlayerPosition;
    search?: string;
}